### ISSUE TITLE
Need way to perform work upon test failure

### ISSUE THREAD

**Issue by newhoggy** [hedgehogqa/haskell-hedgehog#399]
In particular, when a test fails, I'd like a way to perform some `IO` action to read a file and annotate it.

It is insufficient to do the annotation before the assertion because the file is a log file that is growing and will not contain all the information I need, hence the desire to add the annotation at the point of failure.

**Comment by ocharles** [comment]
It sounds like all you want is `onException` from the `exceptions` library. I believe test failure is just raising an exception

**Comment by newhoggy** [comment]
Are exceptions considered part of the API?

**Comment by HuwCampbell** [comment]
```haskell
newtype PropertyT m a =
  PropertyT {
      unPropertyT :: TestT (GenT m) a
    }

newtype TestT m a =
  TestT {
      unTest :: ExceptT Failure (Lazy.WriterT Journal m) a
    }
```

It doesn't use exceptions for failures, though it does catch them. But you can unwrap and rewrap these types as much as you like.

**Comment by newhoggy** [comment]
I got it working:

```
               ┏━━ test/Spec/Network.hs ━━━
            76 ┃ hprop_foo :: Property
            77 ┃ hprop_foo = H.propertyOnce . H.runFinallies $ do
            78 ┃   H.onFailure $ H.noteM_ $ H.readFile "moo.txt"
               ┃   │ Reading file: moo.txt
               ┃   │ hello world
            79 ┃ 
            80 ┃   H.writeFile "moo.txt" "hello world"
               ┃   │ Writing file: moo.txt
            81 ┃ 
            82 ┃   H.failure
               ┃   ^^^^^^^^^
```

**Comment by moodmosaic** [comment]
@newhoggy, great :+1:  Feel free to paste an example here, if that's easy for you, for anyone else that might be looking into this in the future.

**Comment by newhoggy** [comment]
This is the code I used to expose the ability to catch and throw assertions:

https://github.com/input-output-hk/cardano-node/blob/master/hedgehog-extras/src/Hedgehog/Extras/Test/MonadAssertion.hs

The `runFinallies` function is defined here:

https://github.com/input-output-hk/cardano-node/blob/5106f637233a9f2c12faaca571b78378714b4d0a/hedgehog-extras/src/Hedgehog/Extras/Test/Base.hs#L362

And `onFailure` function is defined here:

https://github.com/input-output-hk/cardano-node/blob/5106f637233a9f2c12faaca571b78378714b4d0a/hedgehog-extras/src/Hedgehog/Extras/Test/Base.hs#L349

**Comment by newhoggy** [comment]
I'd love for something like this to be upstreamed into `hedgehog`.

**Comment by uzytkownik** [comment]
> newtype PropertyT m a =
>   PropertyT {
>       unPropertyT :: TestT (GenT m) a
>     }
> 
> newtype TestT m a =
>   TestT {
>       unTest :: ExceptT Failure (Lazy.WriterT Journal m) a
>     }
> 
> It doesn't use exceptions for failures, though it does catch them. But you can unwrap and rewrap these types as much as you like.

It seems to be exposed only in internal module. How stable it is?

**Comment by moodmosaic** [comment]
The internal types have been stable in practice for years and nobody intends to break them. The one known structural concern is that `TestT` uses `Lazy.WriterT`, which can cause space leaks with many annotations in long-running tests.

**Comment by HuwCampbell** [comment]
Should we just move to [Control.Monad.Trans.Writer.CPS](https://hackage.haskell.org/package/transformers-0.6.3.0/docs/Control-Monad-Trans-Writer-CPS.html), which is a drop in replacement, but effectively is a `StateT` under the covers and doesn't suffer from the space leak?

**Comment by moodmosaic** [comment]
Yes, I think moving to Control.Monad.Trans.Writer.CPS makes sense.

One thing to check: the lower bound on `transformers`, as it would require `transformers >= 0.6`. (In practice this probably only affects GHC < 8.6, which seems fine to drop at this point.)
